src/python/pants/backend/javascript/install_node_package.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import itertools
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.javascript import nodejs_project_environment
+from pants.backend.javascript.dependency_inference.rules import rules as dependency_inference_rules
+from pants.backend.javascript.nodejs_project_environment import (
+    NodeJsProjectEnvironment,
+    NodeJsProjectEnvironmentProcess,
+    NodeJSProjectEnvironmentRequest,
+)
+from pants.backend.javascript.package_json import (
+    NodePackageNameField,
+    NodePackageVersionField,
+    PackageJsonSourceField,
+)
+from pants.backend.javascript.subsystems import nodejs
+from pants.backend.javascript.target_types import JSSourceField
+from pants.build_graph.address import Address
+from pants.core.target_types import FileSourceField, ResourceSourceField
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.internals.native_engine import Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import ProcessResult
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import (
+    SourcesField,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+    targets_with_sources_types,
+)
+from pants.engine.unions import UnionMembership, UnionRule
+
+
+@dataclass(frozen=True)
+class InstalledNodePackageRequest:
+    address: Address
+
+
+@dataclass(frozen=True)
+class InstalledNodePackage:
+    project_env: NodeJsProjectEnvironment
+    digest: Digest
+
+    @property
+    def project_dir(self) -> str:
+        return self.project_env.root_dir
+
+
+@dataclass(frozen=True)
+class InstalledNodePackageWithSource(InstalledNodePackage):
+    pass
```

Comment:
Sort of, depends on definition of third party. It also includes in-repo packages that is depended on, because that is the requirement to be able to perform a succesful `npm clean-install`.

I'll admit it is not particularly useful distinction in this pr, but for e.g tests which will be running in `NodeJsProjectEnvironmentProcess`. I believe a prereq for test caching is that only files that are tested are included, i.e you want to `npm clean-install`, but not everytime unrelated test files change.

Reply:
Name change suggestions welcome!